pkg/coscheduling/core/core_test.go

```diff
@@ -0,0 +1,226 @@
+			lastDeniedPG:    denyCache,
+			expectedSuccess: false,
+		},
+		{
+			name:            "pod belongs to a non-existing pg",
+			pod:             st.MakePod().Name("p2").UID("p2").Namespace("ns1").Label(util.PodGroupLabel, "pg-notexisting").Obj(),
+			lastDeniedPG:    newCache(),
+			expectedSuccess: true,
+		},
+		{
+			name: "pod count less than minMember",
+			pod:  st.MakePod().Name("p2").UID("p2").Namespace("ns1").Label(util.PodGroupLabel, "pg1").Obj(),
+			pods: []*corev1.Pod{
+				st.MakePod().Name("pg2-1").UID("pg2-1").Namespace("ns1").Label(util.PodGroupLabel, "pg2").Obj(),
+			},
+			lastDeniedPG:    newCache(),
+			expectedSuccess: false,
+		},
+		{
+			name: "pod count equal minMember",
+			pod:  st.MakePod().Name("p2").UID("p2").Namespace("ns1").Label(util.PodGroupLabel, "pg1").Obj(),
+			pods: []*corev1.Pod{
+				st.MakePod().Name("pg1-1").UID("pg1-1").Namespace("ns1").Label(util.PodGroupLabel, "pg1").Obj(),
+				st.MakePod().Name("pg2-1").UID("pg2-1").Namespace("ns1").Label(util.PodGroupLabel, "pg1").Obj(),
+			},
+			lastDeniedPG:    newCache(),
+			expectedSuccess: true,
+		},
+		{
+			name: "pod count more minMember",
+			pod:  st.MakePod().Name("p2").UID("p2").Namespace("ns1").Label(util.PodGroupLabel, "pg1").Obj(),
+			pods: []*corev1.Pod{
+				st.MakePod().Name("pg1-1").UID("pg1-1").Namespace("ns1").Label(util.PodGroupLabel, "pg1").Obj(),
+				st.MakePod().Name("pg2-1").UID("pg2-1").Namespace("ns1").Label(util.PodGroupLabel, "pg1").Obj(),
+				st.MakePod().Name("pg3-1").UID("pg3-1").Namespace("ns1").Label(util.PodGroupLabel, "pg1").Obj(),
+			},
+			lastDeniedPG:    newCache(),
+			expectedSuccess: true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			cs := clientsetfake.NewSimpleClientset()
+			informerFactory := informers.NewSharedInformerFactory(cs, 0)
+			podInformer := informerFactory.Core().V1().Pods()
+			existingPods, allNodes := makeNodesAndPods(map[string]string{"test": "a"}, 60, 30)
+			snapshot := testutil.NewFakeSharedLister(existingPods, allNodes)
+			pgMgr := &PodGroupManager{pgLister: pgLister, lastDeniedPG: tt.lastDeniedPG, snapshotSharedLister: snapshot, podLister: podInformer.Lister()}
+			informerFactory.Start(ctx.Done())
+			if !clicache.WaitForCacheSync(ctx.Done(), podInformer.Informer().HasSynced) {
+				t.Fatal("WaitForCacheSync failed")
+			}
+			for _, p := range tt.pods {
+				podInformer.Informer().GetStore().Add(p)
+			}
+			err := pgMgr.PreFilter(ctx, tt.pod)
+			if (err == nil) != tt.expectedSuccess {
+				t.Errorf("desire %v, get %v", tt.expectedSuccess, err == nil)
+			}
+		})
+	}
+}
+
+func TestPermit(t *testing.T) {
+	ctx := context.Background()
+	cs := fakepgclientset.NewSimpleClientset()
+
+	pgInformerFactory := pgformers.NewSharedInformerFactory(cs, 0)
+	pgInformer := pgInformerFactory.Scheduling().V1alpha1().PodGroups()
+	pgInformerFactory.Start(ctx.Done())
+
+	pg := integration.MakePG("pg", "ns1", 2)
```

Comment:
Mind moving `MakePG` from `test/integration` to `pkg/util`?

Reply:
I would like keep this, we would only use this for testing.